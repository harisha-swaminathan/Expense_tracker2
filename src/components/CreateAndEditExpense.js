import React, { useState } from "react";
import DatePicker from "react-datepicker";

const CreateAndEditExpense = (props) => {
    const expenseToEdit = props.expenseToEdit || {};    
    const [description, setDescription] =  useState(expenseToEdit.description ? expenseToEdit.description : '');
    const [amount, setAmount] =  useState(expenseToEdit.amount ?  (expenseToEdit.amount/100).toString() : '');
    const [note, setNote] =  useState(expenseToEdit.note ? expenseToEdit.note : '');
    const [createdAt, setCreatedAt] =  useState(expenseToEdit.createdAt || new Date());
    const [error, setError] =  useState('');

    const onDescriptionChange = (e) => {
        setDescription(e.target.value)
    };
    const onAmountChange = (e) => {
        const amount = e.target.value;

        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            setAmount(e.target.value);
        }
    };
    const onNoteChange = (e) => {
        setNote(e.target.value)
    };
    const onDateChange = (createdAt) => {
        if(createdAt) {
            setCreatedAt(createdAt);
        }
    };
    const onExpenseFormSubmit = (e) => {
        e.preventDefault();
        if(!description || !amount) {
            setError('Please provide Description and amount!');
            return;
        }
        setError('');
        props.onSubmit({
            description,
            amount: parseFloat(amount, 10) * 100,
            createdAt: createdAt.valueOf(),
            note
        });
    };
    return (
        <>
            {!!error && <p>{error}</p>}
            <form onSubmit={onExpenseFormSubmit} data-testid="form-element">
                <input
                    type="text"
                    placeholder="Description"
                    onChange={onDescriptionChange}
                    value = {description}
                    autoFocus
                />
                <input
                    type="text"
                    placeholder="Amount"
                    onChange={onAmountChange}
                    value = {amount}
                />
                <DatePicker 
                    selected={createdAt} 
                    showIcon
                    format="dd/MM/yy"
                    onChange={onDateChange} 
                />
                <textarea
                    placeholder="add a note for you expense"
                    onChange={onNoteChange}
                    value = {note}
                />
                <button data-testid="form-submit"> Add Expense</button>
            </form>
        </>
    );
}

export default CreateAndEditExpense;

