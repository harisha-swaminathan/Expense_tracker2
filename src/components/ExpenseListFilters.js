import React, { useState } from 'react';
import {connect} from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setEndDate, setStartDate} from '../actions/filters';
import DatePicker from "react-datepicker";

export const ExpenseListFilter = (props) => {
    const startDate = props.filters.startDate ? props.filters.startDate.toString() : null;
    const endDate = props.filters.endDate ? props.filters.endDate.toString() : null;
    const text = props.filters.text;
    return (
        <>
        <input 
            placeholder='filter by text'
            type="text"
            value={text}
            onChange={(e) => {
                props.setTextFilter(e.target.value)
            }
        }/>
        <select
            data-testid="select-filter"
            value="date"
            onChange={(e) => {
                if(e.target.value === "amount"){
                    props.sortByAmount();
                } else {
                    props.sortByDate();
                }
            }
        }>
            <option value="date">Date</option>
            <option value="amount">Amount</option>
        </select>
        <DatePicker
            selected={startDate}
            showIcon
            onChange={(date) => {
                props.setStartDate(date)
            }}
            selectsStart
            startDate={startDate}
            endDate={endDate}
            dateFormat="MM/dd/yyyy"
            isClearable={true}
        />
        <DatePicker
            selected={endDate}
            showIcon
            onChange={(date) => {
                props.setEndDate(date)
            }}
            selectsEnd
            startDate={startDate}
            endDate={endDate}
            minDate={startDate}
            dateFormat="MM/dd/yyyy"
            isClearable={true}
        />
        </>
    );
};

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
};
const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (value) => {
        dispatch(setTextFilter(value));
    },
    sortByAmount: () => {
        dispatch(sortByAmount());
    },
    sortByDate: () => {
        dispatch(sortByDate());
    },
    setStartDate: (date) => {
        dispatch(setStartDate(date));
    },
    setEndDate: (date) => {
        dispatch(setEndDate(date));
    }
});
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);