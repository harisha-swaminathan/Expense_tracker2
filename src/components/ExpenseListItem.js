import React from "react";
import { Link } from "react-router-dom";

const ExpenseListItem = ({description, amount, createdAt, id, dispatch}) => {
   return  (
        <>
            <Link to={`/edit/${id}`}>
             <h3>{description}</h3>
            </Link> <br/>
            <p>{amount} - {createdAt}</p>
        </>
    );
}


export default ExpenseListItem;