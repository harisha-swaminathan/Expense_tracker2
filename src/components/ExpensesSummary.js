import React from 'react';
import { connect } from 'react-redux';
import getVisibleExpenses from "../selectors/expenses";
import getExpensesTotal from '../selectors/expense-total';
import numeral from 'numeral';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {

    const word = expenseCount === 1 ? 'expense': 'expenses';
    const formatedTotal = numeral((expensesTotal)/100).format('$0,0.00');

    return (
        <>
        viewing {expenseCount} {word} totalling {formatedTotal}!
        </>
    );
};

const mapStateToProps = (state) => {
    
    const expenses = getVisibleExpenses(state.expenses, state.filters);

    return {
        expenseCount: expenses.length,
        expensesTotal: getExpensesTotal(expenses)
    }
    
};

export default connect (mapStateToProps)(ExpensesSummary);