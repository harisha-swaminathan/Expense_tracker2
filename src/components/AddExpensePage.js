import React from 'react';
import CreateAndEditExpense from './CreateAndEditExpense';
import { connect } from "react-redux";
import { addExpense } from "../actions/expenses";
import { useNavigate } from "react-router";

export const AddExpensePage = (props) => {
    const navigator = useNavigate();
    return (
        <div>
            <h1> Add Expense</h1>
            <CreateAndEditExpense 
                onSubmit={(expense) => {
                   props.addExpense(expense)
                    navigator("/");
                }}
            />
        </div>
    );
}

const mapDispatchToProps = (dispatch) => ({
    
    addExpense: ({description, note, amount, createdAt}) => {
        dispatch(addExpense({ 
            description,
            note,
            amount,
            createdAt
        }));
    }
    
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);