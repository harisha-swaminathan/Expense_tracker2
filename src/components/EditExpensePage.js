import React from 'react';
import { useParams } from 'react-router-dom';
import { connect } from 'react-redux';
import CreateAndEditExpense from './CreateAndEditExpense';
import { editExpense , deleteExpense} from '../actions/expenses';
import { useNavigate } from "react-router";

export const EditExpensePage = (props) => {
    const navigator = useNavigate();
    const id = useParams().id;
    const expenseToEdit = props.expenses.find((expense) => expense.id === id);
    const onEditExpense = (updates) => {
        props.editExpense(id, updates);
        navigator("/");
    };
    const onRemove = () => {
        props.deleteExpense(id);
        navigator("/");
    };
    return (
        <div>
            <CreateAndEditExpense expenseToEdit={expenseToEdit} onSubmit={(onEditExpense)}/>
            <button onClick={onRemove}> Remove </button>
        </div>
    )
};

const mapStateToProps = (state) => {
    return {
        expenses: state.expenses
    }
};

const mapDispatchToProps = (dispatch) => {
    return {
        deleteExpense: (id) => {
            dispatch(deleteExpense(id))
        },
        editExpense: (id, updates) => {
            dispatch(editExpense(id, updates))
        }
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);