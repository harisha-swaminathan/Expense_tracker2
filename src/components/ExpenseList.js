import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from "../selectors/expenses";

export const ExpenseList = (props) => {
    return (
        <>
            {
                props.expenses.length ===0 ? (
                    <p> No expenses </p>
                ) : (
                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense}/>
                    })
                )}
        </>
    );
}
const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
};

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);

export default ConnectedExpenseList;