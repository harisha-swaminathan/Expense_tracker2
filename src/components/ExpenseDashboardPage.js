import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilters';

const ExpensesDashboardPage = () => (
    <div>
        
        <ExpenseListFilter/>
        <ExpenseList/>
    </div>
);
export default ExpensesDashboardPage;