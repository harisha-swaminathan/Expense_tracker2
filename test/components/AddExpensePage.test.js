import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

import { BrowserRouter } from 'react-router-dom';
import { AddExpensePage } from "../../src/components/AddExpensePage";

test('should render AddExpensePage', () =>{
    const addExpense = jest.fn();
    const { asFragment } = render( <BrowserRouter> <AddExpensePage addExpense={addExpense}/> </BrowserRouter>);

    expect(asFragment()).toMatchSnapshot();
});
