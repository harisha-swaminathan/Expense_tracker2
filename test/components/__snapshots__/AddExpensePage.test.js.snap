// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpensePage 1`] = `
<DocumentFragment>
   
  <div>
    <h1>
       Add Expense
    </h1>
    <form
      data-testid="form-element"
    >
      <input
        placeholder="Description"
        type="text"
        value=""
      />
      <input
        placeholder="Amount"
        type="text"
        value=""
      />
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container react-datepicker__view-calendar-icon"
        >
          <svg
            class="react-datepicker__calendar-icon "
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M96 32V64H48C21.5 64 0 85.5 0 112v48H448V112c0-26.5-21.5-48-48-48H352V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V64H160V32c0-17.7-14.3-32-32-32S96 14.3 96 32zM448 192H0V464c0 26.5 21.5 48 48 48H400c26.5 0 48-21.5 48-48V192z"
            />
          </svg>
          <input
            class=""
            type="text"
            value="12/23/2024"
          />
        </div>
      </div>
      <textarea
        placeholder="add a note for you expense"
      />
      <button
        data-testid="form-submit"
      >
         Add Expense
      </button>
    </form>
  </div>
   
</DocumentFragment>
`;
